waspc/src/Wasp/Generator/DbGenerator.hs

```diff
@@ -83,42 +83,67 @@ genMigrationsDir spec =
-
   if dbSchemaChecksumFileExists
-    then do
-      dbSchemaFileChecksum <- hexToString <$> checksumFromFilePath dbSchemaFp
-      dbChecksumFileContents <- readFile dbSchemaChecksumFp
-      return $ warnIf (dbSchemaFileChecksum /= dbChecksumFileContents) "Your Prisma schema has changed, you should run `wasp db migrate-dev`."
-    else return $ warnIf entitiesExist "Please run `wasp db migrate-dev` to ensure the local project is fully initialized."
+    then warnIfSchemaDiffersFromChecksum dbSchemaFp dbSchemaChecksumFp
+    else
+      if entitiesExist
+        then warnIfSchemaDiffersFromDb projectRootDir
+        else return Nothing
   where
     dbSchemaFp = SP.fromAbsFile $ projectRootDir </> dbSchemaFileInProjectRootDir
-    dbSchemaChecksumFp = SP.fromAbsFile $ projectRootDir </> dbSchemaChecksumOnLastMigrateFileProjectRootDir
+    dbSchemaChecksumFp = SP.fromAbsFile $ projectRootDir </> dbSchemaChecksumOnLastDbConcurrenceFileProjectRootDir
     entitiesExist = not . null $ getEntities spec
 
-    warnIf :: Bool -> String -> Maybe GeneratorWarning
-    warnIf b msg = if b then Just $ GeneratorNeedsMigrationWarning msg else Nothing
+warnIfSchemaDiffersFromChecksum :: FilePath -> FilePath -> IO (Maybe GeneratorWarning)
+warnIfSchemaDiffersFromChecksum dbSchemaFp dbSchemaChecksumFp = do
+  dbSchemaFileChecksum <- hexToString <$> checksumFromFilePath dbSchemaFp
+  dbChecksumFileContents <- readFile dbSchemaChecksumFp
+  if dbSchemaFileChecksum /= dbChecksumFileContents
+    then return . Just $ GeneratorNeedsMigrationWarning "Your Prisma schema has changed, please run `wasp db migrate-dev` when ready."
+    else return Nothing
+
+warnIfSchemaDiffersFromDb :: Path' Abs (Dir ProjectRootDir) -> IO (Maybe GeneratorWarning)
+warnIfSchemaDiffersFromDb projectRootDir = do
+  -- NOTE: If we wanted to, we could also check that the migrations dir == db,
+  -- but a schema check should handle all most likely cases.
+  schemaMatchesDb <- DbOps.doesSchemaMatchDb projectRootDir
+  case schemaMatchesDb of
+    Just True -> do
+      -- NOTE: This is an optimization, since we know schema == db.
+      -- Writing this file prevents future redundant Prisma checks.
+      DbOps.writeDbSchemaChecksumToFile projectRootDir (SP.castFile dbSchemaChecksumOnLastDbConcurrenceFileProjectRootDir)
+      return Nothing
+    Just False -> return . Just $ GeneratorNeedsMigrationWarning "Your Prisma schema does not match your database, please run `wasp db migrate-dev`."
+    -- NOTE: If there was an error, it could mean we could not connect to the SQLite db, since it does not exist.
+    -- Or it could mean their DATABASE_URL is wrong, or database is down, or any other number of causes.
+    -- In any case, migrating will either solve it (in the SQLite case), or allow Prisma to give them enough info to troubleshoot.
+    Nothing -> return . Just $ GeneratorNeedsMigrationWarning "Please run `wasp db migrate-dev` to ensure the local project is fully initialized."
```

Comment:
Maybe we could provide just a bit of context additionally, right now it sounds like running bd migrate-dev is certainly solution while we actually don't know. Maybe say failed to perform the check and that their best bet is to run db migrate-dev to fix this or to get more info on the source of the error.

Reply:
I tried going back and forth with that, had a hard time making it sound right though :) Will give it more thought on how to convey it.

Reply:
Let me know if I can help with this somehow.